rfc/java11.md

```diff
@@ -0,0 +1,164 @@
+Java 11 can use the `java8` variant by specifying `imageVariant: slim-java8` (or
+`ubuntu-java8`) in their Halyard config file.
+
+The release notes will prominently mention this fact. It will also be mentioned
+in-channel in the release announcement to `#spinnaker-releases` and in the
+release announcement email. We will strongly encourage people who opt for the
+`java8` variants to contact `sig-platform@spinnaker.io` or `#sig-platform` to
+discuss their concerns, which will allow us to adjust the timeline as needed.
+
+Separately, postings to `#dev` will encourage developers to start running their
+development environments under Java 11. Explanations of how to do so will be
+posted to the channel.
+
+If issues are found in Spinnaker that are caused by running under a Java 11
+runtime, we will switch the containers back to Java 8 and shift the timeline
+back one release. Since we're providing the `java8` variants, we will not adjust
+the timeline for any other reason.
+
+### 1.19 (early March)
+
+_All_ containers built by the build process will now use the Java 11 JRE. As
+before, the `java8` variants will offer a way to opt out of this change.
+
+Other than that, everything from the previous section applies to this release as
+well.
+
+### 1.20 (early May)
+
+For this release, we will do two things:
+
+1.  Compile with JDK 11 using the `-source 8 -target 8` flags. This produces
+    builds that are still compatible with Java 8, and means the build will still
+    fail if anyone tries to use Java 11 APIs or language features.
+2.  Remove the `java8` container variants
+
+If the compiler change causes issues, we can revert that change and push that
+out to the next release.
+
+If we know for a fact (from previous discussions) that removing the `java8`
+variants will cause problems for some customers, or if a customer complains
+after the release, we can put back the `java8` containers and delay this part
+(but keep the compiler change) until the next release.
+
+### 1.21 (early July)
+
+Change `-source 8 -target 8` to `-source 11 -target11`. At this point,
+developers are able to commit Java 11 code to the repositories. Undoing this
+change is possible, but will require some work. As such, this should only be
+done after
+
+1.  the `java8` containers have been removed for a while without complaint
+2.  we don't know of any other non-container customers still running on Java 8
+
+## Prior Art and Alternatives
+
+The general approach was modelled after how Google migrates its monorepo to new
+Java versions:
+
+1.  switch the JRE to the new version
+2.  use the new JDK's `javac` with `-target 8 -source 8`
+3.  remove those flags
+
+Between each step is a period of time to evaluate breakages and potentially roll
+back.
+
+## Known Unknowns
+
+On the JRE side, there are probably changes to performance, memory usage, and
+anything else that could be affected by a major runtime change like this.
```

Comment:
Do you have any plans for how to measure this other than users reporting perf issues? Would it be helpful to take a baseline against, say, the integration tests, and report any noticeable difference as the timeline progresses? This might give us an opportunity to set some users at ease when reading release notes by saying something like "In our tests, memory and CPU consumption was N% above/below our baseline". Also, if we see anything majorly concerning, we can address it prior to a release.

Reply:
I think instrumenting the integration tests in a way that would give us that data would be a very significant amount of work. We can't really just use a wall clock because the random latency from talking to cloud providers is almost certainly going to hide any changes from the JRE. So we'd have to stick a profiling tool in there somewhere and measure CPU usage and frankly I'm not sure the workloads are similar enough to give any useful data anyway. So I don't think the return on the investment would really be worth it.